lib/wibox/layout/ratio.lua

```diff
@@ -0,0 +1,339 @@
+
+        pos = pos + space + spacing
+
+        if (self.dir == "y" and pos-spacing >= height) or
+            (self.dir ~= "y" and pos-spacing >= width) then
+            break
+        end
+    end
+
+    return result
+end
+
+--- Increase the ratio of "widget"
+-- If the increment produce an invalid ratio (not between 0 and 1), the method
+-- do nothing.
+-- @tparam number index The widget index to change
+-- @tparam number increment An floating point value between -1 and 1 where the
+--   end result is within 0 and 1
+function ratio:inc_ratio(index, increment)
+    if #self.widgets ==  1 or (not index) or increment < -1 or increment > 1 then return end
+
+    local widget = self.widgets[index]
+
+    assert(self._ratios[index])
+
+    self:set_ratio(index, self._ratios[index] + increment)
+end
+
+--- Increment the ratio of the first instance of `widget`
+-- If the increment produce an invalid ratio (not between 0 and 1), the method
+-- do nothing.
+-- @param widget The widget to ajust
+-- @tparam number increment An floating point value between -1 and 1 where the
+--   end result is within 0 and 1
+function ratio:inc_widget_ratio(widget, increment)
+    if not widget or not increment then return end
+
+    local index = self:index(widget)
+
+    self:inc_ratio(index, increment)
+end
+
+--- Set the ratio of the widget at position `index`
+-- @tparam number index The widget to change or its index
+-- @tparam number percent An floating point value between 0 and 1
+function ratio:set_ratio(index, percent)
+    if not percent or #self.widgets ==  1 or not index or index > #self.widgets
+        or percent < 0 or percent > 1 then 
+        return
+    end
+
+    if not self._ratios[index] then
+        normalize(self)
+    end
+
+    local old = self._ratios[index]
+
+    -- Remove what has to be cleared from all widget
+    local delta = ( (percent-old) / (#self.widgets-1) )
+
+    for k, v in pairs(self.widgets) do
+        self._ratios[k] = self._ratios[k] - delta
+    end
+
+    -- Set the new ratio
+    self._ratios[index] = percent
+
+    -- As some widgets may now have a slightly negative ratio, normalize again
+    normalize(self)
+
+    self:emit_signal("widget::layout_changed")
+end
+
+--- Set the ratio of "widget"
+-- @param widget The widget to ajust
+-- @tparam number percent An floating point value between 0 and 1
+function ratio:set_widget_ratio(widget, percent)
+    local index = self:index(widget)
+
+    self:set_ratio(index, percent)
+end
+
+--- Update all widgets to match a set of a ratio
+-- @tparam number index The widget to change or its index
+-- @tparam number before The sum of the ratio before the widget
+-- @tparam number itself The ratio for "widget"
+-- @tparam number after The sum of the ratio after the widget
+function ratio:ajust_ratio(index, before, itself, after)
+    if not self.widgets[index] or not before or not itself or not after then
+        return
+    end
```

Comment:
In `:set_ratio` you check for things like `<0` or `>1`. Should that be done here as well?